### ISSUE TITLE
Please provide slimv-compatible keyboard bindings

### ISSUE THREAD

**Issue by phmarek** [l04m33/vlime#19]
That would allow much easier transitioning/testing of your plugin.

I'd hope for some config setting that determines whether traditional vlime or slimv compatible ones are installed.

Thank you so much for considering this!

**Comment by l04m33** [comment]
Thanks for the suggestion!

I think this depends on how you define "compatible". Key mappings can be easily added to invoke similar features in Vlime. Yet Vlime was not designed to be compatible with Slimv, so these features, although similar to the ones found in Slimv, will not behave the exact same way as in Slimv.

If Slimv users can tolerate some subtle differences in Vlime behavior, I can certainly add Slimv-like key mappings. But I'm not quite sure about this.

I'd recommend new users to go through the Vlime tutorial (`:help vlime-tutor`), and then remap the keys to their heart's content (`:help vlime-mappings-remap`).

**Comment by phmarek** [comment]
Hi Kay,
thanks for the quick answer!

> I'd recommend new users to go through the Vlime tutorial (:help vlime-tutor), and then remap the keys to their heart's content (:help vlime-mappings-remap).

Yes, of course that's possible. But why have that work (badly) duplicated, when you could do it once?
(I'm using the long keybindings, by the way.)

One interesting thing that is possible with Slimv is that there is a function to store forms in a register *and* push them on at the same time, and another one that takes a register name, pushes the contents from the register and *then* the selected forms.
This means I have `<F2>` bound to the former, and `<F3>` to the latter; with `<F2>` I can choose a test form to evaluate, and on pressing `<F3>` the form the cursor is on gets evaluated (typically a `(DEFUN ...)`) and the test expression is sent on, too; this way a single keypress is enough to see the result of changes in the REPL. I guess a similar thing will be possible in Vlime, too.

**Comment by l04m33** [comment]
> But why have that work (badly) duplicated, when you could do it once?

You've got a point there. As I stated before, Vlime cannot behave the same way as Slimv does, so we'll have to induce approximation and/or compromises when implementing the Slimv-like key mappings. I think the crucial part would be determining what approximation/compromises should there be.

I looked at the [documentation](https://github.com/kovisoft/slimv/blob/bcfaa598752ff15cf4c02f6a34886dd2a7fe5e64/doc/slimv.txt#L653) for Slimv and assessed the compatibility in each group of commands (✔ means Vlime has all the features in that group, and those features are generally compatible with the ones found in Slimv):

* The Slimv menu: Vlime doesn't have a global menu.
* Edit commands: Vlime doesn't have the "Close Form" feature, nor the built-in paredit mode. The "Close Form" feature may be added in the future though.
* Evaluation commands: Vlime has all these features, but doesn't use any Vim register to store Lisp expressions.
* Debug commands: ✔
* Compile commands: Vlime has all these features, but doesn't use any Vim register to store Lisp expressions.
* Cross reference commands: ✔
* Profile commands: Vlime has no profiling support. May be added in the future.
* Documentation commands: Vlime doesn't know how to search the Hyperspec.
* REPL commands: ✔
* REPL menu commands: These commands doesn't make sense in Vlime, due to the way Vlime's REPL works.

In the short term, I think we can have this to-do list, to approach a good-enough Slimv-compatible mode:

- [ ] Implement the "Close Form feature.
- [ ] Implement profiling support.
- [ ] Handle the use of Vim registers.
- [ ] Provide Slimv-like key mappings.

So what do you think?

**Comment by l04m33** [comment]
As for storing forms in registers, you can use the existing Vlime API to build your own command for now. For example, a function similar to the "Eval Defun" command in Slimv:

```vim
function! EvalAndPutTopLevelExpr()
    execute 'let' '@'.v:register '=' 'vlime#ui#CurTopExpr()'
    execute 'call' 'vlime#plugin#SendToREPL(@'.v:register.')'
endfunction
```

`vlime#ui#CurTopExpr()` and `vlime#plugin#SendToREPL()` are part of the public API, you can look them up in Vlime's documentation.

**Comment by phmarek** [comment]
Well, perhaps your to-do list should be the other way around...
Having compatible key mappings would allow people to easily start testing it, and that means more potential contributors.

That Slimv uses registers is simply because these are available as a prefix; for my use-case any other internal variable would do as well.

**Comment by l04m33** [comment]
> Well, perhaps your to-do list should be the other way around...
> Having compatible key mappings would allow people to easily start testing it, and that means more potential contributors.

Good point. I'll try to come up with a set of close-enough key mappings, and see how it goes.

**Comment by l04m33** [comment]
I've added a set of Slimv-like mappings in commit 60234b2. Please have a try. Just add `let g:vlime_overlay = "slimv"` in your `vimrc` (see `:help vlime-mappings-overlays`).

**Comment by phmarek** [comment]
Hi,

I started testing a bit; you did some great work!

Would you prefer individual issues for the small problems I've found, or should I just drop them all here?

**Comment by l04m33** [comment]
> Would you prefer individual issues for the small problems I've found, or should I just drop them all here?

Thanks for asking :) I think both ways are fine.
